server/src/lib/config.rs

```diff
@@ -144,6 +148,76 @@ impl BlockDevice {
 
                 Ok((be, child))
             }
+            "crucible" => {
+                let targets: Vec<SocketAddrV4> = self
+                    .options
+                    .get("targets")
+                    .ok_or_else(|| {
+                        ParseError::KeyNotFound(
+                            "targets".to_string(),
+                            "options".to_string(),
+                        )
+                    })?
+                    .as_array()
+                    .ok_or_else(|| {
+                        ParseError::AsError(
+                            "targets".to_string(),
+                            "as_array".to_string(),
+                        )
+                    })?
+                    .to_vec()
+                    .iter()
+                    .map(|x| {
+                        x.as_str()
+                            .ok_or_else(|| {
+                                ParseError::AsError(
+                                    "x".to_string(),
+                                    "as_str".to_string(),
+                                )
+                            })?
+                            .parse()
+                            .map_err(|e| {
+                                ParseError::AsError(
+                                    "x".to_string(),
+                                    format!("parse(): {:?}", e),
+                                )
+                            })
+                    })
+                    .collect::<Result<Vec<SocketAddrV4>, ParseError>>()?;
+
+                let read_only: bool = || -> Option<bool> {
+                    self.options.get("readonly")?.as_str()?.parse().ok()
+                }()
+                .unwrap_or(false);
+
+                let key: Option<Result<String, ParseError>> = self
+                    .options
+                    .get("key")
+                    .map(|x| -> Result<String, ParseError> {
+                        Ok(x.as_str()
+                            .ok_or_else(|| {
+                                ParseError::AsError(
+                                    "x".to_string(),
+                                    "as_str".to_string(),
+                                )
+                            })?
+                            .to_string())
+                    });
+
+                let key: Option<String> =
+                    if let Some(key) = key { Some(key?) } else { None };
```

Comment:
You can replace this by adding `.map_or(Ok(None), |r| r.map(Some))?;` to the above statement.